### PR TITLE
influxdata: add more Tokenizer methods

### PR DESCRIPTION
**Pull request by rogpeppe**
The testing is still on the light side, because we can't introduce the
whole corpus testing without also introducing value parsing,
which we'll do in the next PR to keep this one relatively smaller.

### CODE REVIEW

File: influxdata/tokenizer.go [hunk 1]
```diff
@@ -13,7 +13,25 @@ const (
...
+	measurementChars      = newByteSet(",").union(whitespace).invert()
+	measurementEscapes    = newEscaper(" ,")
+	tagKeyChars           = newByteSet(",=").union(whitespace).invert()
+	tagKeyEscapes         = newEscaper(",= ")
+	tagValChars           = newByteSet(",").union(whitespace).invert()
+	tagValEscapes         = newEscaper(", =")
+	fieldSeparatorSpace   = newByteSet(" \t\r\f")
+	fieldKeyChars         = tagKeyChars
+	fieldKeyEscapes       = tagKeyEscapes
+	fieldStringValChars   = newByteSet(`"`).invert()
```
Comment (philjb) [inline influxdata/tokenizer.go:26]: Single quotes are forbidden in field value strings: https://docs.influxdata.com/influxdb/v2.0/reference/syntax/line-protocol/#quotes

File: influxdata/tokenizer.go [hunk 2]
```diff
@@ -47,20 +69,389 @@ type Tokenizer struct {
...
+	case '"':
+		// Skip leading quote.
+		t.advance(1)
+		fieldVal = t.takeEsc(fieldStringValChars, &fieldStringValEscapes.revTable)
+		fieldKind = String
+		if !t.ensure(1) {
+			return nil, Unknown, nil, t.syntaxErrorf("expected closing quote for string field value, found end of input")
+		}
+		if t.at(0) != '"' {
+			// This shouldn't be able to happen, as all characters are allowed in a string.
```
Comment (philjb) [inline influxdata/tokenizer.go:272]: I don't believe single quotes are allowed nor unescaped backslashes. Dunno what your corpus says.

**Review by philjb — APPROVED** [review]
I put some minor improvements and suggestions into this branch `rog-006-tokenizer-methods-pjb` for your consideration - mostly around comments, error messages, but also that bit about which chars are allowed in tag values.

**Comment by philjb** [comment]
The build failed from a static check failure. I believe the commit message isn't conventional either.

File: influxdata/tokenizer.go [hunk 3]
```diff
@@ -13,7 +13,25 @@ const (
...
 
-// Tokenizer implements low level parsing of a line-protocol message.
+var (
+	measurementChars      = newByteSet(",").union(whitespace).invert()
+	measurementEscapes    = newEscaper(" ,")
+	tagKeyChars           = newByteSet(",=").union(whitespace).invert()
+	tagKeyEscapes         = newEscaper(",= ")
+	tagValChars           = newByteSet(",").union(whitespace).invert()
+	tagValEscapes         = newEscaper(", =")
+	fieldSeparatorSpace   = newByteSet(" \t\r\f")
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply influxdata/tokenizer.go:23]: > this is so sad that it's not simpler.

+100

Unfortunately we have what we have, and different parsers have different ideas of what consists of white space.
All parsers seem to accept tab; not all accept `\r` or `\f` characters, but I think it's worth treating them as whitespace for now. I intend to instrument the data gathering to see whether anyone is actually using these different space characters in practice.

The docs are vague at best and flat-out wrong at worst, I'm afraid.

File: influxdata/tokenizer.go [hunk 4]
```diff
@@ -13,7 +13,25 @@ const (
 	minRead = minGrow / 2
 )
 
-// Tokenizer implements low level parsing of a line-protocol message.
+var (
+	measurementChars      = newByteSet(",").union(whitespace).invert()
+	measurementEscapes    = newEscaper(" ,")
+	tagKeyChars           = newByteSet(",=").union(whitespace).invert()
+	tagKeyEscapes         = newEscaper(",= ")
+	tagValChars           = newByteSet(",").union(whitespace).invert()
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply influxdata/tokenizer.go:21]: Yes, you're right - `=` isn't accepted unescaped as part of a tag value, although if there _is_ an unescaped `=` character, it's a syntax error, not just the end of the tag section.

File: influxdata/tokenizer.go [hunk 5]
```diff
@@ -13,7 +13,25 @@ const (
...
+	tagValEscapes         = newEscaper(", =")
+	fieldSeparatorSpace   = newByteSet(" \t\r\f")
+	fieldKeyChars         = tagKeyChars
+	fieldKeyEscapes       = tagKeyEscapes
+	fieldStringValChars   = newByteSet(`"`).invert()
+	fieldStringValEscapes = newEscaper(`\"`)
+	fieldValChars         = newByteSet(",").union(whitespace).invert()
+	timeChars             = newByteSet("-0123456789")
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply influxdata/tokenizer.go:29]: Yes, that would be nice but we're not trying to redesign lp here. That's a vague glimmer of hope shining in the far distance :)

File: influxdata/tokenizer.go (hunk 2 above)
Comment (rogpeppe) [reply influxdata/tokenizer.go:272]: Single quotes are allowed in strings (the doc is misleading here) - you just can't use single quotes to quote a string.
You can have unescaped backslashes - an backslash followed by an unknown escape character represents exactly a backslash followed by that character. So `\q` encodes to just itself. Not that `fieldStringValChars` is _exactly_ everything except `"`.

File: influxdata/tokenizer.go [hunk 6]
```diff
@@ -47,20 +69,389 @@ type Tokenizer struct {
...
+		case '\n':
+			t.advance(1)
+		default:
+			return
+		}
+	}
+}
+
+func (t *Tokenizer) advanceToSection(section Section) (_ok bool, _err error) {
+	defer func() {
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply influxdata/tokenizer.go:387]: just a debugging remnant. will remove.

File: influxdata/tokenizer.go (hunk 1 above)
Comment (rogpeppe) [reply influxdata/tokenizer.go:26]: The doc is misleading in this respect. See my comment elsewhere.

**Pull request merged**